master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/StateMachine.java

```diff
@@ -72,94 +61,7 @@
-         */
-        @Override
-        public void cleanupOldSnapshots(SnapshotRetentionPolicy snapshotRetentionPolicy)
-            throws IOException {
-          if (snapshotRetentionPolicy != null
-              && snapshotRetentionPolicy.getNumSnapshotsRetained() > 0) {
-            List<SingleFileSnapshotInfo> allSnapshotFiles = new ArrayList<>();
-            List<SingleFileSnapshotInfo> allMD5Files = new ArrayList<>();
-            try (DirectoryStream<Path> stream =
-                Files.newDirectoryStream(SimpleStateMachineStorageUtil.getSmDir(this).toPath())) {
-              for (Path path : stream) {
-                if (filePatternMatches(SNAPSHOT_REGEX, allSnapshotFiles, path)) {
-                  continue;
-                } else {
-                  filePatternMatches(MD5_REGEX, allMD5Files, path);
-                }
-              }
-            }
-            // first step, cleanup old snapshot and md5 file
-            SingleFileSnapshotInfo snapshotInfo =
-                cleanupOldFiles(
-                    allSnapshotFiles,
-                    snapshotRetentionPolicy.getNumSnapshotsRetained(),
-                    false,
-                    null);
-            // second step, cleanup only old md5 file
-            cleanupOldFiles(
-                allMD5Files, snapshotRetentionPolicy.getNumSnapshotsRetained(), true, snapshotInfo);
-          }
-        }
-
-        private boolean filePatternMatches(
-            Pattern pattern, List<SingleFileSnapshotInfo> result, Path filePath) {
-          Matcher md5Matcher = pattern.matcher(filePath.getFileName().toString());
-          if (md5Matcher.matches()) {
-            final long endIndex = Long.parseLong(md5Matcher.group(2));
-            final long term = Long.parseLong(md5Matcher.group(1));
-            final FileInfo fileInfo = new FileInfo(filePath, null);
-            result.add(new SingleFileSnapshotInfo(fileInfo, term, endIndex));
-            return true;
-          }
-          return false;
-        }
-
-        private SingleFileSnapshotInfo cleanupOldFiles(
-            List<SingleFileSnapshotInfo> inputFiles,
-            int retainedNum,
-            boolean onlyCleanupMD5Files,
-            SingleFileSnapshotInfo snapshotInfo) {
-          SingleFileSnapshotInfo result = null;
-          if (inputFiles.size() > retainedNum) {
-            inputFiles.sort(new RatisSnapshotFileComparator());
-            List<SingleFileSnapshotInfo> filesToBeCleaned =
-                inputFiles.subList(retainedNum, inputFiles.size());
-            result = filesToBeCleaned.get(0);
-            for (SingleFileSnapshotInfo fileInfo : filesToBeCleaned) {
-              if ((null != snapshotInfo && (fileInfo.getIndex() >= snapshotInfo.getIndex())
-                  || (onlyCleanupMD5Files && null == snapshotInfo))) {
-                continue;
-              }
-              File file = fileInfo.getFile().getPath().toFile();
-              if (onlyCleanupMD5Files) {
```

Comment:
BTW, is this old logic correct? IIUC when `onlyCleanupMD5Files` is true, we should delete the md5 file instead of the original file. cc @FMX @RexXiong

Reply:
the old logic is wrong here,(compared to `ratis-2.5.1`), we also need clean up the md5 files if the corresponding snapshot file does not exist.

https://github.com/apache/ratis/blob/77640a3a1d012d0225d26a5a52b38fb80bb9fe44/ratis-server/src/main/java/org/apache/ratis/statemachine/impl/SimpleStateMachineStorage.java#L129

Reply:
Let me give you one example (follow the old logic), we take 5 snapshots and set the numSnapshotsRetained to 3.

![image](https://github.com/apache/incubator-celeborn/assets/95013770/98ad79d3-ecc6-4f2a-ae5d-030663f666e3)

Reply:
Hi @xleoken 
After I checked the old logic, there should be md5 files be left over and they will be cleaned. As we can see that the old logic delete snapshot files in two steps.
1. Find all snapshot files and delete them with the md5 files.
2. Find all md5 files without the corresponding snapshot files and delete them.
```java
try (DirectoryStream<Path> stream =
                Files.newDirectoryStream(SimpleStateMachineStorageUtil.getSmDir(this).toPath())) {
              for (Path path : stream) {
                if (filePatternMatches(SNAPSHOT_REGEX, allSnapshotFiles, path)) {
                  continue;
                } else {
                  filePatternMatches(MD5_REGEX, allMD5Files, path);
                }
              }
            }


            // first step, cleanup old snapshot and md5 file
            SingleFileSnapshotInfo snapshotInfo =
                cleanupOldFiles(
                    allSnapshotFiles,
                    snapshotRetentionPolicy.getNumSnapshotsRetained(),
                    false,
                    null);
            // second step, cleanup only old md5 file
            cleanupOldFiles(
                allMD5Files, snapshotRetentionPolicy.getNumSnapshotsRetained(), true, snapshotInfo);
```

Have you found out why ratis will left with md5 files only?

Reply:
> Have you found out why ratis will left with md5 files only?

Hi @FMX, these md5 files are generated in `MasterStateMachineSuiteJ#testSnapshotCleanup` test case.

From the ratis project, we can see that clean up the md5 files if the corresponding snapshot file does not exist.

![image](https://github.com/apache/incubator-celeborn/assets/95013770/57fa83e5-becb-48a3-b819-9ed9c6516b0b)

Reply:
I add `Arrays.stream(remainingFiles).forEach(item -> System.out.println(item));` to `MasterStateMachineSuiteJ#testSnapshotCleanup`.
![image](https://github.com/apache/incubator-celeborn/assets/95013770/63f6e3af-2ee2-4712-9e6d-db3e64333a1b)

Reply:
It's by design to keep the number of `snapshotRetentionPolicy.getNumSnapshotsRetained()` files. Ratis won't create MD5 only.
Check the codes below.
```java
private SingleFileSnapshotInfo cleanupOldFiles(
            List<SingleFileSnapshotInfo> inputFiles,
            int retainedNum,
            boolean onlyCleanupMD5Files,
            SingleFileSnapshotInfo snapshotInfo) {
          SingleFileSnapshotInfo result = null;
          if (inputFiles.size() > retainedNum) {
            inputFiles.sort(new RatisSnapshotFileComparator());
            List<SingleFileSnapshotInfo> filesToBeCleaned =
                inputFiles.subList(retainedNum, inputFiles.size());
            result = filesToBeCleaned.get(0);
            for (SingleFileSnapshotInfo fileInfo : filesToBeCleaned) {
              if ((null != snapshotInfo && (fileInfo.getIndex() >= snapshotInfo.getIndex())
                  || (onlyCleanupMD5Files && null == snapshotInfo))) {
                continue;
              }
              File file = fileInfo.getFile().getPath().toFile();
              if (onlyCleanupMD5Files) {
                LOG.info("Deleting old md5 file at {}.", file.getAbsolutePath());
                FileUtils.deleteFileQuietly(file);
              } else {
                File md5File = new File(file.getAbsolutePath() + MD5FileUtil.MD5_SUFFIX);
                LOG.info(
                    "Deleting old snapshot at {}, md5 file at {}.",
                    file.getAbsolutePath(),
                    md5File.getAbsolutePath());
                FileUtils.deleteFileQuietly(file);
                FileUtils.deleteFileQuietly(md5File);
              }
            }
          }
          return result;
        }
```

Reply:
Due to the Ratis community is not very active, the old logic has been verified for a while. Can you verify this PR in your test environment?

Reply:
> Due to the Ratis community is not very active, the old logic has been verified for a while. Can you verify this PR in your test environment?

@FMX @waitinfuture take a look again when free.

Yeah, I run a celeborn cluster(3 master with 1 worker), and run some shuffle examples.

And let `raft.server.snapshot.retention.file.num` to `3`, after tests, obtained the expected results.

**Conf**
```
celeborn.master.ha.ratis.raft.server.storage.dir                           /opt/data/celeborn_ratis/
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.enabled         true
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold       1
celeborn.master.ha.ratis.raft.server.snapshot.retention.file.num           3
```

**Snapshots**
```
root@kaaa:/opt/data# 
root@kaaa:/opt/data# curl 172.17.0.3:9098/masterGroupInfo
====================== Master Group INFO ==============================
group id: c5196f6d-2c34-3ed3-8b8a-47bede733167
leader info: 1(kaaa.hadoop.com:9872)

[server {
  id: "2"
  address: "172.17.0.3:9872"
  clientAddress: "172.17.0.3:9097"
  startupRole: FOLLOWER
}
commitIndex: 82
, server {
  id: "1"
  address: "kaaa.hadoop.com:9872"
  clientAddress: "172.17.0.2:9097"
  startupRole: FOLLOWER
}
commitIndex: 82
, server {
  id: "3"
  address: "172.17.0.4:9872"
  clientAddress: "172.17.0.4:9097"
  startupRole: FOLLOWER
}
commitIndex: 82
]
root@kaaa:/opt/data# 
root@kaaa:/opt/data# 
root@kaaa:/opt/data# ll /opt/data/celeborn_ratis/c5196f6d-2c34-3ed3-8b8a-47bede733167/sm
total 24
drwxr-xr-x 2 root root 144 Oct 19 07:50 ./
drwxr-xr-x 5 root root  61 Oct 19 07:31 ../
-rw-r--r-- 1 root root 175 Oct 19 07:50 snapshot.1_80
-rw-r--r-- 1 root root  48 Oct 19 07:50 snapshot.1_80.md5
-rw-r--r-- 1 root root 175 Oct 19 07:50 snapshot.1_82
-rw-r--r-- 1 root root  48 Oct 19 07:50 snapshot.1_82.md5
-rw-r--r-- 1 root root 175 Oct 19 07:50 snapshot.1_84
-rw-r--r-- 1 root root  48 Oct 19 07:50 snapshot.1_84.md5
root@kaaa:/opt/data# 
root@kaaa:/opt/data# 
root@kaaa:/opt/data# 
```

Reply:
@xleoken Thanks for the test, this patch will be merged into main and branch-0.3.